okhttp/src/main/java/okhttp3/CertificatePinner.kt

```diff
@@ -0,0 +1,339 @@
+      val x509Certificate = peerCertificates[c] as X509Certificate
+
+      // Lazily compute the hashes for each certificate.
+      var sha1: ByteString? = null
+      var sha256: ByteString? = null
+
+      for (pin in pins) {
+        when (pin.hashAlgorithm) {
+          "sha256/" -> {
+            if (sha256 == null) sha256 = sha256(x509Certificate)
+            if (pin.hash == sha256) return  // Success!
+          }
+          "sha1/" -> {
+            if (sha1 == null) sha1 = sha1(x509Certificate)
+            if (pin.hash == sha1) return  // Success!
+          }
+          else -> throw AssertionError("unsupported hashAlgorithm: ${pin.hashAlgorithm}")
+        }
+      }
+    }
+
+    // If we couldn't find a matching pin, format a nice exception.
+    val message = buildString {
+      append("Certificate pinning failure!")
+      append("\n  Peer certificate chain:")
+      for (c in 0 until peerCertificates.size) {
+        val x509Certificate = peerCertificates[c] as X509Certificate
+        append("\n    ")
+        append(pin(x509Certificate))
+        append(": ")
+        append(x509Certificate.subjectDN.name)
+      }
+      append("\n  Pinned certificates for ")
+      append(hostname)
+      append(":")
+      for (pin in pins) {
+        append("\n    ")
+        append(pin)
+      }
+    }
+    throw SSLPeerUnverifiedException(message)
+  }
+
+  @Deprecated(
+      "replaced with {@link #check(String, List)}.",
+      ReplaceWith("check(hostname, peerCertificates.toList())")
+  )
+  @Throws(SSLPeerUnverifiedException::class)
+  inline fun check(hostname: String, vararg peerCertificates: Certificate) {
+    check(hostname, peerCertificates.toList())
+  }
+
+  /**
+   * Returns list of matching certificates' pins for the hostname. Returns an empty list if the
+   * hostname does not have pinned certificates.
+   */
+  internal fun findMatchingPins(hostname: String): List<Pin> {
+    var result: List<Pin> = emptyList()
+    for (pin in pins) {
+      if (pin.matches(hostname)) {
+        if (result.isEmpty()) result = mutableListOf()
+        (result as MutableList<Pin>).add(pin)
+      }
+    }
+    return result
+  }
+
+  /** Returns a certificate pinner that uses `certificateChainCleaner`.  */
+  internal fun withCertificateChainCleaner(
+    certificateChainCleaner: CertificateChainCleaner?
+  ): CertificatePinner {
+    return if (this.certificateChainCleaner == certificateChainCleaner) {
+      this
+    } else {
+      CertificatePinner(pins, certificateChainCleaner)
+    }
+  }
+
+  override fun equals(other: Any?): Boolean {
+    if (other === this) return true
+    return other is CertificatePinner
+        && certificateChainCleaner == other.certificateChainCleaner
+        && pins == other.pins
+  }
+
+  override fun hashCode(): Int {
+    var result = Objects.hashCode(certificateChainCleaner)
+    result = 31 * result + pins.hashCode()
+    return result
+  }
+
+  internal class Pin(
```

Comment:
For an internal class is it worth making this a data class?  moving more of the logic into a companion object method?

Reply:
I like it as a data class. For follow-up though.